File: 121-best-time-to-buy-and-sell-stock.go
func maxProfit(prices []int) int {
    maxProfit := 0
    bestLowPrice := prices[0]

    for _, price := range prices[1:] {
        maxProfit = max(maxProfit, price - bestLowPrice)
        bestLowPrice = min(bestLowPrice, price)
    }

    return maxProfit
}
func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}
func max(a, b int) int {
    if a >= b {
        return a
    }
    return b
}

File: 121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxProfit = 0
        bestLowPrice = prices[0]

        for price in prices[1:]:
            maxProfit = max(maxProfit, price - bestLowPrice)
            bestLowPrice = min(bestLowPrice, price)
        return maxProfit

  
